### ISSUE TITLE
Hushboard doesn't work with Wayland

### ISSUE THREAD

**Issue by stuartlangridge** [stuartlangridge/hushboard#13]
Not sure if it's the same error, but in my case it only seems to work on certain applications (Wayland vs Xorg). We should probably adapt the software a little to get the keystrokes from both environments.

To reproduce the issue, you can use my setup:

- Arch Linux
- Sway
- Firefox
- Alacritty

Then open an alacritty terminal (wayland) and the Firefox browser. Start to type on the terminal and then try on Firefox.
Result: keystrokes on Alacritty (Wayland) are not detected, thus the microphone is not muted. Keystrokes on Firefox (Xorg) are detected and the microphone is correctly muted.

_Originally posted by @denysvitali in https://github.com/stuartlangridge/hushboard/issues/6#issuecomment-757682837_

**Comment by stuartlangridge** [comment]
@denysvitali I can certainly confirm that Hushboard doesn't work in Wayland; I don't believe it _can_ work in Wayland, and it doesn't attempt to. Am I right in thinking that you are running some sort of a mixture of Wayland and X, so that Hushboard _does_ work when you press keys in X apps?

**Comment by denysvitali** [comment]
That's correct! Wayland input is not detected, but Xorg is. By looking at the source code, this is an expected behavior since only Xorg is handled.

I've also tried to fix the issue myself by adding a Wayland client using pywayland, but I didn't succeed: I don't think I'm able to fix the issue myself, but at least I understood the cause 👍

**Comment by stuartlangridge** [comment]
Wayland being architected differently, I am not sure it is possible at all for one application to see keypresses of another (and note that Hushboard doesn't need to see what the keypresses _are_, just that there were some). So I don't think this can be fixed, but I may be wrong on that; I'm not running Wayland myself so I can't tell. I'll leave this issue open for now rather than WONTFIXing it because if I'm wrong perhaps someone else can contribute a Wayland patch!

**Comment by bluelightzero** [comment]
Wayland is not comparable to X.org.

X is the server, but Wayland is just a protocol that servers (known as compositors under wayland) can use to talk to clients. There is no wayland process.

When using a wayland compositor, you can start up something called XWayland which runs what is basically an X Server and sends any X commands from X clients to the current wayland compositor. Keypress listeners can ask Xwayland for key presses, But if an application is a wayland client then it will skip past this.

If you want to support wayland, you should be looking at the APIs for each popular desktop environment (gnome, kde, ect) to see if they allow it. They may share a common library for handling this which is ideal. I don't know of one. They might not even be one.

Will get back to you, because I really want this on wayland.

**Comment by bluelightzero** [comment]
This might actually help make this work: https://www.codedbearder.com/posts/writing-keyboard-remapper-wayland/

Apparently a work around is to get exclusive access to a keyboard, then create a virtual keyboard that echos the input.

**Comment by nekohayo** [comment]
Maybe the only way you could achieve this is to somehow plug onto whatever mechanism GNOME Settings / GSD / libinput uses to disable touchpads while the user is typing? At least that works on Wayland, because it's a built-in core feature, but you'd have to talk to GNOME folks to figure out if there might be any way to tap into that. The only hint I have is that the feature is globally controlled by this hidden dconf key: `org.gnome.desktop.peripherals.touchpad disable-while-typing true`

**Comment by bluelightzero** [comment]
This was a cool concept, but AI noise suppression is much better now and just works.

**Comment by swapdisk** [comment]
I was just today looking for a solution to do exactly what Hushboard does, but then noticed it doesn't work with Wayland. So I did what anybody living in the year 2026 would do, I used AI to vibe code my own. 

It uses `evdev` for the keyboard event monitoring, so by working at the kernel input device level, it completely gets around annoying Wayland. I also had to overcome Gnome deliberately dropping the `input` group from desktop session processes, but got around that doing using `sg input`.

Anyway, I just wanted to leave a bread crumb here for anybody coming upon this issue. You can find the Python code and detailed README at [swapdisk/magic-mute](https://github.com/swapdisk/magic-mute).
